chain-vote/src/private_voting/unit_vector_zkp.rs

```diff
@@ -0,0 +1,475 @@
+        };
+
+        Proof {
+            ibas: first_announcement_vec,
+            ds: poly_coeff_enc,
+            zwvs: randomness_response_vec,
+            r: response,
+        }
+    }
+
+    /// Verify a unit vector proof. The verifier checks that the plaintexts encrypted in `ciphertexts`,
+    /// under `public_key` is a unit vector.
+    /// Common Reference String (`crs`): Pedersen Commitment Key
+    /// Statement: public key `pk`, and ciphertexts `encrypting_vote.ciphertexts`
+    /// C_0=Enc_pk(r_0; v_0), ..., C_{m-1}=Enc_pk(r_{m-1}; v_{m-1})
+    ///
+    /// Description of the verification procedure available in Figure 9.
+    pub(crate) fn verify(
+        &self,
+        crs: &Crs,
+        public_key: &PublicKey,
+        ciphertexts: &[Ciphertext],
+    ) -> bool {
+        let ck = CommitmentKey::from(crs.clone());
+        let ciphertexts = Ptp::new(ciphertexts.to_vec(), Ciphertext::zero);
+        let bits = ciphertexts.bits();
+        let mut cc = ChallengeContext::new(&ck, public_key, ciphertexts.as_ref());
+        let cy = cc.first_challenge(&self.ibas);
+        let cx = cc.second_challenge(&self.ds);
+
+        if self.ibas.len() != bits {
+            return false;
+        }
+
+        if self.zwvs.len() != bits {
+            return false;
+        }
+
+        self.verify_statements(public_key, &ck, &ciphertexts, &cx, &cy)
+    }
+
+    /// Final verification of the proof, that we compute in a single vartime multiscalar
+    /// multiplication.
+    #[cfg(feature = "ristretto255")]
+    fn verify_statements(
+        &self,
+        public_key: &PublicKey,
+        commitment_key: &CommitmentKey,
+        ciphertexts: &Ptp<Ciphertext>,
+        challenge_x: &Scalar,
+        challenge_y: &Scalar,
+    ) -> bool {
+        let bits = ciphertexts.bits();
+        let length = ciphertexts.len();
+        let cx_pow = challenge_x.power(bits);
+
+        let powers_cx = challenge_x.exp_iter();
+        let powers_cy = challenge_y.exp_iter();
+
+        let powers_z_iterator = powers_z_encs_iter(&self.zwvs, challenge_x, &(bits as u32));
+
+        let zero = public_key.encrypt_with_r(&Scalar::zero(), &self.r);
+
+        let mega_check = GroupElement::multiscalar_multiplication(
+            self.zwvs
+                .iter()
+                .map(|zwv| zwv.z)
+                .chain(self.zwvs.iter().map(|zwv| zwv.w))
+                .chain(iter::repeat(challenge_x.negate()).take(bits))
+                .chain(iter::repeat(Scalar::one().negate()).take(bits))
+                .chain(self.zwvs.iter().map(|zwv| zwv.v))
+                .chain(self.zwvs.iter().map(|zwv| zwv.z - challenge_x))
+                .chain(iter::repeat(Scalar::one().negate()).take(bits))
+                .chain(powers_cy.take(length).map(|s| s * cx_pow))
+                .chain(powers_cy.take(length).map(|s| s * cx_pow))
+                .chain(powers_cy.take(length).map(|s| s))
+                .chain(powers_cx.take(bits).map(|s| s))
```

Comment:
What's up with `map(|s| s)`?

Reply:
woops, first implementation was over `RistrettoPoint`s, so it was `map(|s| s.0)`. Removed 👍